conjure-go-client/httpclient/internal/request_retrier.go

```diff
@@ -65,119 +55,73 @@ func (r *RequestRetrier) attemptsRemaining() bool {
-	if !retryFn() {
-		return "", false
-	}
-	return r.currentURI, r.isRelocatedURI(r.currentURI)
-}
 
-func (r *RequestRetrier) getRetryFn(resp *http.Response, respErr error) func() bool {
-	errCode, _ := StatusCodeFromError(respErr)
-	if retryOther, _ := isThrottleResponse(resp, errCode); retryOther {
-		// 429: throttle
-		// Immediately backoff and select the next URI.
-		// TODO(whickman): use the retry-after header once #81 is resolved
-		return r.nextURIAndBackoff
-	} else if isUnavailableResponse(resp, errCode) {
-		// 503: go to next node
-		return r.nextURIOrBackoff
-	} else if shouldTryOther, otherURI := isRetryOtherResponse(resp, respErr, errCode); shouldTryOther {
-		// 307 or 308: go to next node, or particular node if provided.
-		if otherURI != nil {
-			return func() bool {
-				r.setURIAndResetBackoff(otherURI)
-				return true
-			}
-		}
-		return r.nextURIOrBackoff
-	} else if errCode >= http.StatusBadRequest && errCode < http.StatusInternalServerError {
-		return nil
-	} else if resp == nil {
-		// if we get a nil response, we can assume there is a problem with host and can move on to the next.
-		return r.nextURIOrBackoff
+	if r.isSuccess(resp) {
+		return false, nil
 	}
-	return nil
-}
 
-func (r *RequestRetrier) setURIAndResetBackoff(otherURI *url.URL) {
-	// If the URI returned by relocation header is a relative path
-	// We will resolve it with the current URI
-	if !otherURI.IsAbs() {
-		if currentURI := parseLocationURL(r.currentURI); currentURI != nil {
-			otherURI = currentURI.ResolveReference(otherURI)
-		}
+	if r.isNonRetryableClientError(resp, err) {
+		return false, nil
 	}
-	nextURI := otherURI.String()
-	r.relocatedURIs[otherURI.String()] = struct{}{}
-	r.retrier.Reset()
-	r.currentURI = nextURI
-}
 
-// If lastURI was already marked failed, we perform a backoff as determined by the retrier before returning the next URI and its offset.
-// Otherwise, we add lastURI to failedURIs and return the next URI and its offset immediately.
-func (r *RequestRetrier) nextURIOrBackoff() bool {
-	_, performBackoff := r.failedURIs[r.currentURI]
-	r.markFailedAndMoveToNextURI()
-	// If the URI has failed before, perform a backoff
-	if performBackoff || len(r.uris) == 1 {
-		return r.retrier.Next()
+	// handle redirects
+	if tryOther, otherURI := isRetryOtherResponse(resp, err); tryOther {
+		return true, otherURI
 	}
-	return true
-}
 
-// Marks the current URI as failed, gets the next URI, and performs a backoff as determined by the retrier.
-func (r *RequestRetrier) nextURIAndBackoff() bool {
-	r.markFailedAndMoveToNextURI()
-	return r.retrier.Next()
-}
+	// don't retry mesh uris
+	if r.isMeshURI(resp) {
+		return false, nil
+	}
 
-func (r *RequestRetrier) markFailedAndMoveToNextURI() {
-	r.failedURIs[r.currentURI] = struct{}{}
-	nextURIOffset := (r.offset + 1) % len(r.uris)
-	nextURI := r.uris[nextURIOffset]
-	r.currentURI = nextURI
-	r.offset = nextURIOffset
+	// retry with backoff
+	return r.retrier.Next(), nil
```

Comment:
nit: I would separate backoff behavior into a different interface given the weird way we use the underlying `Retrier` interface is not intuitive.